### ISSUE TITLE
Metabase DB Migration H2 > MySQL Script Errors

### ISSUE THREAD

**Issue by erix920** [metabase/metabase#7596]
When trying to migrate the default H2 database to MySQL I ran into some errors.

First - columns names that are reserved in SQL weren't surrounded by the ` character. The error I got was: check the manual that corresponds to your MySQL server version for the right syntax to use near ‘admin BIT(1) NOT NULL, user_id INT NOT NULL,

To fix this I had to set automigrate to false, copy script and update any columns that were named row or admin to be `row` and `admin`.

Second - a few tables weren't fully qualified with the database. They were: 
- update report_dashboard set public_perms = 2 where public_perms = 1;
- update metabase_database set is_sample = true where name = ‘Sample Dataset’;
- update metabase_database set is_full_sync = true;

Original post can be found here: http://discourse.metabase.com/t/error-migrating-from-h2-to-mysql/1599

**Comment by salsakran** [comment]
what version of metabase were you running?

**Comment by erix920** [comment]
It was .29

**Comment by senior** [comment]
@erix920 I'm not able to reproduce this issue. I'm wondering if it's related to the version you are migrating from, MySQL version or maybe there's some improperly escaped data that is causing issues.

What version of MySQL are you attempting to migrate to? Also just checking, it's MySQL and not MariaDB.

You mentioned that you are using v0.29, I assume that's the version you're using to run the migration. Is that also the version that was previously ran on that H2 database that you are trying to migrate?

When I run the `load_from_h2` command, I do get properly escaped table names. Are you able to get the SQL that is being executed from the MySQL log and paste it here? Maybe there are some characters in the data that's causing bad SQL to be created or something, not sure, just trying to find out what's different from my setup and yours.

**Comment by rchar01** [comment]
I wanted to migrate Metabase (0.29.3) to the MariaDB (10.2.10) and I had a similar error:
```
You have an error in your SQL syntax; check the manual that corresponds to your MariaDB 
server version for the right syntax to use near 
'rows INT NULL, description TEXT NULL, entity_name VARCHAR(254) NULL, entity_type' at line 1
```
so in `metabase.metabase_table`, `rows` column should be surrounded by the ` character.

**Comment by senior** [comment]
@rchar01 The error you are seeing is related to a difference between MySQL and MariaDB. Maria diverged from MySQL to make `rows` a reserved word. Currently we only support Maria as long as it's compatible with MySQL. Could you try migrating to MySQL instead? The ticket being used for Maria support is https://github.com/metabase/metabase/issues/2581. Some people have had luck generating SQL and quoting the rows column (info [here](https://github.com/metabase/metabase/issues/2581#issuecomment-330324268)) but since MySQL is what we officially support, it will probably be easier to switch to that.

**Comment by erix920** [comment]
@senior It is MySQL version 8.0.11. I can attempt to reproduce with the specifics if needed. I'm not sure if it matters but I may have been trying to migrate the original v29 H2 database to MySQL using the v30 Metabase jar. I had downloaded both jars so I may have mixed them up.

**Comment by senior** [comment]
@erix920 Thanks for the info, I think the MySQL version was the key piece of info missing. I was using 5.5.56 and 5.7.18 and it works fine there. If I try migrating to MySQL 8.0.11 I get the same error you did. Thanks for the info! I should be able to put together a fix now that I can recreate the error.

**Comment by rchar01** [comment]
Maybe every column name (for MySQL/MariaDB) should be surrounded by the ` character.
What do you think about it?

**Comment by senior** [comment]
The issue here is MySQL 8 added a new reserved word, `admin`. Metabase has a table with a column name `admin` that is not being escaped by Liquibase (the library we use for our database migrations). It looks like the MySQL folks added it as a reserved word in 8, but they are making it non-reserved when 8.0.12 is released (according to the reference docs [here](https://dev.mysql.com/doc/refman/8.0/en/keywords.html)).

This issue doesn't just affect migrating from H2 to MySQL 8. Metabase will fail to start for anyone wanting to use MySQL 8.0.X for the Metabase application database.

I see where the change needs to be made in the Liquibase code and will work on getting a fix for it there.

**Comment by senior** [comment]
@rchar01 for queries we do and we have tests that ensure we are quoting everything. This SQL is being generated by Liquibase though, so we don't control that. Even if we would quote every column, we'd want tests and other validation that we work on Maria before we stated that we officially supported it.

**Comment by senior** [comment]
Created a JIRA issue for liquibase [CORE-3235](https://liquibase.jira.com/browse/CORE-3235). Opened a pull request against the 3.5.x branch https://github.com/liquibase/liquibase/pull/773 that should fix the problem.

**Comment by senior** [comment]
liquibase/liquibase#773 has been merged and should be included in the next release in the 3.5 series, which is 3.5.6. Once that's out we should upgrade and the issue will be fixed.

**Comment by senior** [comment]
The fix for this has been merged with https://github.com/metabase/metabase/pull/8047 and will be included in the next Metabase release, 0.30.0, which should be out soon. More fixes were needed in addition to the Liquibase fix, which are detailed in the commit messages of https://github.com/metabase/metabase/pull/8047.
